src/mono/mono/mini/monovm.c

```diff
@@ -276,3 +277,89 @@ monovm_shutdown (int *latchedExitCode)
 
 	return 0;
 }
+
+static int
+monovm_create_delegate_impl (const char* assemblyName, const char* typeName, const char *methodName, void **delegate);
+	
+
+int
+monovm_create_delegate (const char *assemblyName, const char *typeName, const char *methodName, void **delegate)
+{
+	int result;
+	/* monovm_create_delegate may be called instead of monovm_execute_assembly.  Initialize the
+	 * runtime if it isn't already. */
+	if (!mono_get_root_domain())
+		mini_init (assemblyName, "v4.0.30319");
+	MONO_ENTER_GC_UNSAFE;
+	result = monovm_create_delegate_impl (assemblyName, typeName, methodName, delegate);
+	MONO_EXIT_GC_UNSAFE;
+	return result;
+}
+
+int
+monovm_create_delegate_impl (const char* assemblyName, const char* typeName, const char *methodName, void **delegate)
+{
+	// Load an assembly and a type and a method from the type, then return a pointer to a native
+	// callable version of the method.  See CorHost2::CreateDelegate
+	//
+	// We have to do this in general, but the CoreCLR hostpolicy only calls this with a handful
+	// of methods from the [CoreLib]Internal.Runtime.InteropServices.ComponentActivator
+	// class. See hostpolicy.cpp get_delegate()
+
+	ERROR_DECL (error);
+
+	const int failure = 0x80004005; /* E_FAIL */
+
+	if (!delegate)
+		return failure;
+	*delegate = NULL;
+
+	MonoAssemblyLoadContext *alc = mono_alc_get_default ();
+
+	MonoImage *image;
+	if (!strcmp (MONO_ASSEMBLY_CORLIB_NAME, assemblyName)) {
+		image = mono_defaults.corlib;
+	} else {
+		MonoAssemblyName aname = {0};
+		aname.name = assemblyName;
+		MonoAssemblyByNameRequest req;
+		mono_assembly_request_prepare_byname (&req, alc);
+		MonoImageOpenStatus status = MONO_IMAGE_OK;
+		MonoAssembly *assm = mono_assembly_request_byname (&aname, &req, &status);
+
+		if (!assm || status != MONO_IMAGE_OK)
+			return failure;
+		image = assm->image;
+	}
+
+	MonoType *t = mono_reflection_type_from_name_checked ((char*)typeName, alc, image, error);
+	goto_if_nok (error, fail);
+
+	g_assert (t);
+	MonoClass *klass = mono_class_from_mono_type_internal (t);
+	
+
+	MonoMethod *method = mono_class_get_method_from_name_checked (klass, methodName, -1, 0, error);
+	goto_if_nok (error, fail);
+
+	if (!mono_method_has_unmanaged_callers_only_attribute (method)) {
```

Comment:
I think this is a fine limitation on Mono &ndash; especially since it should be the canonical way to use the API in both runtimes. We should update the tutorial (https://docs.microsoft.com/dotnet/core/tutorials/netcore-hosting) and also update the doc somewhere about the fact that Mono requires the `UnmanagedCallersOnly` on exports.

/cc @elinor-fung @vitek-karas

Reply:
> I think this is a fine limitation on Mono – especially since it should be the canonical way to use the API in both runtimes. We should update the tutorial (https://docs.microsoft.com/dotnet/core/tutorials/netcore-hosting) and also update the doc somewhere about the fact that Mono requires the `UnmanagedCallersOnly` on exports.
> 
> /cc @elinor-fung @vitek-karas


@AaronRobinsonMSFT  I don't think there's anything user-visible here - `monovm_create_delegate` is the implementation of `hostfxr_get_runtime_delegate` - which is only called on a few well-known internal entrypoints (https://github.com/dotnet/runtime/blob/fcee44ab15d287ce45272b65fc217a239ec5927c/src/native/corehost/hostpolicy/hostpolicy.cpp#L527-L538) and the two that Mono cares about are both `[UnmanagedCallersOnly]`.

Reply:
> I don't think there's anything user-visible here - monovm_create_delegate is the implementation of hostfxr_get_runtime_delegate - which is only called on a few well-known internal entrypoints

I think it is user visible since calling `hostfxr_get_runtime_delegate` when mono is the target runtime has limitations. With this change the delegate is always be ignored or must be `UNMANAGEDCALLERSONLY_METHOD` which means the API has limitations depending on which runtime is being used.

I also could be missing your point, but the fact that DNNE users can't use the delegate approach in both runtimes seems user facing enough that it should be documented at the public API level.